mono/mini/mini.c

```diff
@@ -2393,6 +2393,14 @@ static void mono_bb_ordering (MonoCompile *cfg)
 	cfg->max_block_num = cfg->num_bblocks;
 
 	df_visit (cfg->bb_entry, &dfn, cfg->bblocks);
+
+#if defined(__GNUC__) && __GNUC__ == 7 && defined(__x86_64__)
```

Comment:
If it is really the processor bug..can we remove the gnuc && gnuc==7 condition, and just leave defined (__x86__64__) || defined(_M_AMD64) (Windows also).

Reply:
Looks ok, but its more likely to be a compiler bug than a hw one.

Reply:
So far I haven't seen it with other compilers, but, should it happen, I've added the assertion below in this case.

On the other hand, the memory barrier doesn't cost much here (I would almost argue nothing, as this isn't hot code) and it would give us some additional safety.

Reply:
Also might be interesting to test on 32bit x86 on the same machine..

Reply:
Regarding the assert, I do not trust asserts to find bugs reliably. They help but they also can be either insufficient or themselves buggy.

Reply:
@vargaz it could an issue specific to gcc 7 in that it generates a code sequence that triggers the bug.

Reply:
We know that `-O2` triggers it. I got all the options it turns on from gcc docs and I'm turning off the optimizations one by one to see when I can build without crash. Right now the smallest set I got is:

```
-O2 -fno-ipa-cp -fno-ipa-bit-cp -fno-ipa-vrp -fno-ipa-sra -fno-ipa-icf -fno-ipa-ra -fno-align-functions -fno-align-jumps -fno-align-loops -fno-align-labels -fno-peephole2 -fno-expensive-optimizations -fno-reorder-functions -fno-reorder-blocks-and-partition -fno-store-merging -fno-strict-aliasing -fno-code-hoisting -fno-rerun-cse-after-loop -fno-cse-follow-jumps -fno-cse-skip-blocks -fno-schedule-insns -fno-schedule-insns2 -fno-sched-interblock -fno-sched-spec -fno-thread-jumps
```

Reply:
Second set of flags which works is:

```
-O2  -fno-ipa-cp -fno-ipa-bit-cp -fno-ipa-vrp -fno-ipa-sra -fno-ipa-icf -fno-ipa-ra -fno-align-functions -fno-align-jumps -fno-align-loops -fno-align-labels -fno-peephole2 -fno-expensive-optimizations -fno-reorder-functions -fno-reorder-blocks-and-partition -fno-store-merging -fno-strict-alia
sing -fno-code-hoisting -fno-rerun-cse-after-loop -fno-cse-follow-jumps -fno-cse-skip-blocks -fno-schedule-insns -fno-schedule-insns2 -fno-sched-interblock -fno-sched-spec -fno-isolate-erroneous-paths-dereference
```

(only the last option is different)

Third set of flags which works is:

```
-O2  -fno-ipa-cp -fno-ipa-bit-cp -fno-ipa-vrp -fno-ipa-sra -fno-ipa-icf -fno-ipa-ra -fno-align-functions -fno-align-jumps -fno-align-loops -fno-align-labels -fno-peephole2 -fno-expensive-optimizations -fno-reorder-functions -fno-reorder-blocks-and-partition -fno-store-merging -fno-strict-alia
sing -fno-code-hoisting -fno-rerun-cse-after-loop -fno-cse-follow-jumps -fno-cse-skip-blocks -fno-schedule-insns -fno-schedule-insns2 -fno-sched-interblock -fno-sched-spec -fno-gcse
```

(again, the last flag differs)